### ISSUE TITLE
[Unity-iOS] NotSupportException

### ISSUE THREAD

**Issue by yjlintw** [Unity-Technologies/ml-agents#54]
I tried the iOS with TensorFlowSharp example, and got the following error when deploy to a mobile device. It works fine on Unity Editor.

Here is the Error Message:
``` (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)

NotSupportedException: To marshal a managed method, please add an attribute named 'MonoPInvokeCallback' to the method definition.
  at TensorFlow.TFTensor..ctor (System.Int32 value) [0x00000] in <00000000000000000000000000000000>:0 
  at TensorFlow.TFTensor.op_Implicit (System.Int32 value) [0x00000] in <00000000000000000000000000000000>:0 
  at ExampleCommon.ImageUtil.ConstructGraphToNormalizeImage (TensorFlow.TFGraph& graph, TensorFlow.TFOutput& input, TensorFlow.TFOutput& output, TensorFlow.TFDataType destinationDataType) [0x00000] in <00000000000000000000000000000000>:0 
  at ExampleCommon.ImageUtil.CreateTensorFromImageFile (System.String file, TensorFlow.TFDataType destinationDataType) [0x00000] in <00000000000000000000000000000000>:0 
  at TensorTest.Test () [0x00000] in <00000000000000000000000000000000>:0 
 
(Filename: currently not available on il2cpp Line: -1)

Setting up 1 worker threads for Enlighten.
  Thread -> id: 16f54b000 -> priority: 1 
```

Here is my source code (grab from TensorFlowSharp Example)
``` csharp
using System.IO;
using UnityEngine;
using TensorFlow;

namespace ExampleCommon
{
	public static class ImageUtil
	{
		// Convert the image in filename to a Tensor suitable as input to the Inception model.
		public static TFTensor CreateTensorFromImageFile (string file, TFDataType destinationDataType = TFDataType.Float)
		{
			byte[] contents;
			if (Application.platform == RuntimePlatform.Android)
			{
				WWW reader = new WWW(file);
				while (!reader.isDone) { }

				contents = reader.bytes;
			} else {
				contents = File.ReadAllBytes(file);
			}

			// var contents = File.ReadAllBytes (file);

			// DecodeJpeg uses a scalar String-valued tensor as input.
			var tensor = TFTensor.CreateString (contents);

			TFGraph graph;
			TFOutput input, output;

			// Construct a graph to normalize the image
			ConstructGraphToNormalizeImage (out graph, out input, out output, destinationDataType);

			// Execute that graph to normalize this one image
			using (var session = new TFSession (graph)) {
				var normalized = session.Run (
						 inputs: new [] { input },
						 inputValues: new [] { tensor },
						 outputs: new [] { output });

				return normalized [0];
			}
		}

		// The inception model takes as input the image described by a Tensor in a very
		// specific normalized format (a particular image size, shape of the input tensor,
		// normalized pixel values etc.).
		//
		// This function constructs a graph of TensorFlow operations which takes as
		// input a JPEG-encoded string and returns a tensor suitable as input to the
		// inception model.
		private static void ConstructGraphToNormalizeImage (out TFGraph graph, out TFOutput input, out TFOutput output, TFDataType destinationDataType = TFDataType.Float)
		{
			// Some constants specific to the pre-trained model at:
			// https://storage.googleapis.com/download.tensorflow.org/models/inception5h.zip
			//
			// - The model was trained after with images scaled to 224x224 pixels.
			// - The colors, represented as R, G, B in 1-byte each were converted to
			//   float using (value - Mean)/Scale.

			const int W = 224;
			const int H = 224;
			const float Mean = 127.5f;
			const float Scale = 127.5f;

			graph = new TFGraph ();
			input = graph.Placeholder (TFDataType.String);

			output = graph.Cast (graph.Div (
				x: graph.Sub (
					x: graph.ResizeBilinear (
						images: graph.ExpandDims (
							input: graph.Cast (
								graph.DecodeJpeg (contents: input, channels: 3), DstT: TFDataType.Float),
							dim: graph.Const (0, "make_batch")),
						size: graph.Const (new int [] { W, H }, "size")),
					y: graph.Const (Mean, "mean")),
				y: graph.Const (Scale, "scale")), destinationDataType);
		}
	}
}
```

**Comment by vincentpierre** [comment]
Hi @yjlintw thank you for your post. I have a few questions : 
 * You said it works in the editor, does it work in standalone and/or on Android ?
 * Have you successfully put the Balance Ball using Internal Brain example onto an iOS device ?
 * Have you followed the [additional instructions for iOS](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Using-TensorFlow-Sharp-in-Unity-(Experimental).md#ios-additional-instructions-for-building) ? 
 
The code you attached is not familiar to me, I hope we can help.

**Comment by yjlintw** [comment]
1. It works on Standalone, or under any platform when I execute it through pressing the play button in the Unity Editor. It does not work when I build for iOS or Android, but the error are different.
2. I haven't tried that, I'll try it later. I didn't try it because I am only trying to do image recognition.
3. I did, and I assume it should be like this?
<img width="665" alt="screen shot 2017-09-29 at 5 59 06 pm" src="https://user-images.githubusercontent.com/1399376/31037626-f3fad680-a53f-11e7-9c3f-be00155d47cd.png">

The code is from here: https://github.com/migueldeicaza/TensorFlowSharp/blob/master/Examples/ExampleCommon/ImageUtil.cs

Thank you.

**Comment by vincentpierre** [comment]
The error you are getting might be because some operators are not present in the tensorflow libraries for mobile. (mobile phones are lighter so not all operators can be imported). A simple way to test this would be to try to build a simpler app for mobile (Balance Ball or something even simpler) to see if it works.    
What does the error look like on Android ?

**Comment by yjlintw** [comment]
The error on Android comes from ```graph.DecodeJpeg```, which I read another post on stackoverflow saying that Android does not support this function.

Maybe this is the reason that cause the error I saw on iOS?

**Comment by vincentpierre** [comment]
I think you found the reason then. Let me know if you can run the balance ball, it would confirm this. Do you think you could build an image recognition graph that do not use this function ?

**Comment by yjlintw** [comment]
Hmm, I don't know. It will require more research. I'm relatively new to TensorFlow.

I'll let you know when I test the balance ball later today. Thanks.

**Comment by luvial** [comment]
Hello !
I have the same issue on android, is there a way to work around the fact that we can't use DecodeJpeg ?

The model asks to be fed with 4 dimensional input, how could I make it work with Raw image data (byte[])
I tried to modify the "ConstructGraphToNormalizeImage " function from above, but I'm not skilled enough to fix this...

Thank you.

**Comment by mmattar** [comment]
Closing this out as we have since released v0.3 of ML-Agents which improves the Android support. If you continue running into issues with v0.3, please create a new issue.
